tracee-rules/engine/engine.go

```diff
@@ -108,17 +124,81 @@ func (engine *Engine) consumeSources(done <-chan bool) {
 						}
 					}
 				}
+				engine.signaturesMutex.RUnlock()
 				engine.inputs.Tracee = nil
 			} else if event != nil {
+				engine.signaturesMutex.RLock()
 				for _, s := range engine.signaturesIndex[types.SignatureEventSelector{Source: "tracee", Name: event.(tracee.Event).EventName}] {
 					engine.signatures[s] <- event
 				}
 				for _, s := range engine.signaturesIndex[types.SignatureEventSelector{Source: "tracee", Name: "*"}] {
 					engine.signatures[s] <- event
 				}
+				engine.signaturesMutex.RUnlock()
 			}
 		case <-done:
 			return
 		}
 	}
 }
+
+//LoadSignature will store in Engine data structures the given signature and activate its handling business logics.
+// It will return the signature ID as well as error.
+func (engine *Engine) LoadSignature(signature types.Signature) (string, error) {
+	selectedEvents, err := signature.GetSelectedEvents()
+	if err != nil {
+		return "", fmt.Errorf("failed to store signature: %w", err)
+	}
+
+	// insert in engine.signatures map
+	metadata, _ := signature.GetMetadata()
+	engine.signaturesMutex.Lock()
+	defer engine.signaturesMutex.Unlock()
+	if engine.signatures[signature] != nil {
+		// signature already exists
+		return metadata.ID, nil
+	}
+	c := make(chan types.Event)
+	engine.signatures[signature] = c
+
+	// insert in engine.signaturesIndex map
+	for _, selectedEvent := range selectedEvents {
+		engine.signaturesIndex[selectedEvent] = append(engine.signaturesIndex[selectedEvent], signature)
```

Comment:
Seems to be missing some logic as in NewEngine():
```
			if es.Name == "" {
				es.Name = "*"
			}
			if es.Source == "" {
				engine.logger.Printf("signature %s doesn't declare an input source", meta.Name)
			} else {
				engine.signaturesIndex[es] = append(engine.signaturesIndex[es], sig)
			}
```

Reply:
@yanivagman do we still want to add to the engine.signatures map if es.Source == "" ?

Reply:
I don't see a reason to do this. also today we reject it